File: version3/rust/fp.rs
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

use xxx::big;
use xxx::big::BIG;
use xxx::dbig::DBIG;
use xxx::rom;
use arch::Chunk;

//#[cfg(D32)]
use arch::DChunk;

#[derive(Copy, Clone)]
pub struct FP {
 	x:BIG
}

pub const NOT_SPECIAL:usize =0;
pub const PSEUDO_MERSENNE:usize=1;
pub const MONTGOMERY_FRIENDLY:usize=2;
pub const GENERALISED_MERSENNE:usize=3;

pub const MODBITS:usize = @NBT@; /* Number of bits in Modulus */
pub const MOD8: usize = @M8@;  /* Modulus mod 8 */
pub const MODTYPE:usize=@MT@;

pub const FEXCESS:Chunk = ((1 as Chunk)<<(big::BASEBITS*(big::NLEN)-MODBITS));
pub const OMASK:Chunk = (-1)<<(MODBITS%big::BASEBITS);
pub const TBITS:usize=MODBITS%big::BASEBITS; // Number of active bits in top word
pub const TMASK:Chunk=(1<<TBITS)-1;


impl FP {

/* calculate Field Excess */
    pub fn excess(a:&BIG) -> Chunk {
        return (a.w[big::NLEN-1]&OMASK)>>(MODBITS%big::BASEBITS)
    }

//#[cfg(D32)]
    pub fn pexceed(a: &BIG,b: &BIG) -> bool {
        let ea=FP::excess(a);
        let eb=FP::excess(b);        
        if ((ea+1) as DChunk)*((eb+1) as DChunk) > FEXCESS as DChunk {return true}
        return false
    }

//#[cfg(D32)]
    pub fn sexceed(a: &BIG) -> bool {
        let ea=FP::excess(a);
        if ((ea+1) as DChunk)*((ea+1) as DChunk) > FEXCESS as DChunk {return true}
        return false
    }
/*
#[cfg(D64)]
    pub fn pexceed(a: &BIG,b: &BIG) -> bool {
        let ea=FP::excess(a);
        let eb=FP::excess(b);        
        if (ea+1) > FEXCESS/(eb+1) {return true}
        return false
    }

#[cfg(D64)]
    pub fn sexceed(a: &BIG) -> bool {
        let ea=FP::excess(a);
        if (ea+1) > FEXCESS/(ea+1) {return true}
        return false
    }    
*/
/* Constructors */
	pub fn new() -> FP {
		FP {
				x: BIG::new()
		}
	}

	pub fn new_int(a:isize) -> FP {
		let mut f=FP::new(); 
		f.x.inc(a);
		f.nres();
		return f;		
	}

	pub fn new_copy(y:&FP) -> FP {
		let mut f=FP::new(); 
		f.x.copy(&(y.x));
		return f;
	}

	pub fn new_big(y:&BIG) -> FP {
		let mut f=FP::new(); 
		f.x.copy(y);
        f.nres();
		return f;		
	}

    pub fn nres(&mut self) {
        if MODTYPE != PSEUDO_MERSENNE && MODTYPE != GENERALISED_MERSENNE {
   			let p = BIG::new_ints(&rom::MODULUS);        	
            let mut d=DBIG::new_scopy(&(self.x));
            d.shl(big::NLEN*(big::BASEBITS as usize));
            self.x.copy(&d.dmod(&p));
        }
    }

/* convert back to regular form */
    pub fn redc(&mut self) -> BIG {
        if MODTYPE != PSEUDO_MERSENNE && MODTYPE != GENERALISED_MERSENNE {
            let mut d=DBIG::new_scopy(&(self.x));
            return FP::modulo(&mut d);
        } else {
            let r=BIG::new_copy(&(self.x));
            return r;
        }
    }


/* reduce a DBIG to a BIG using the appropriate form of the modulus */
/* dd */
    pub fn modulo(d: &mut DBIG) -> BIG {
    
        if MODTYPE==PSEUDO_MERSENNE {
            let mut b=BIG::new();
            let mut t=d.split(MODBITS);
            b.dcopy(&d);
            let v=t.pmul(rom::MCONST as isize);
            let tw=t.w[big::NLEN-1];
            t.w[big::NLEN-1] &= TMASK;
            t.w[0]+=rom::MCONST*((tw>>TBITS)+(v<<(big::BASEBITS-TBITS)));
    
            b.add(&t);
            b.norm();
            return b;
        }
    
        if MODTYPE==MONTGOMERY_FRIENDLY
        {
            let mut b=BIG::new();
            for i in 0 ..big::NLEN {
                let x=d.w[i];

                let tuple=BIG::muladd(x,rom::MCONST-1,x,d.w[big::NLEN+i-1]);
                d.w[big::NLEN+i]+=tuple.0; d.w[big::NLEN+i-1]=tuple.1;
            }
    
            b.zero();
    
            for i in 0 ..big::NLEN {
                b.w[i]=d.w[big::NLEN+i];
            }
            b.norm();
            return b;
        }
            
        if MODTYPE==GENERALISED_MERSENNE
        { // GoldiLocks Only
            let mut b=BIG::new();            
            let t=d.split(MODBITS);
            let rm2=(MODBITS/2) as usize;
            b.dcopy(&d);
            b.add(&t);
            let mut dd=DBIG::new_scopy(&t);
            dd.shl(rm2);
            
            let mut tt=dd.split(MODBITS);
            let lo=BIG::new_dcopy(&dd);
            b.add(&tt);
            b.add(&lo);
            b.norm();
            tt.shl(rm2);
            b.add(&tt);
            
            let carry=b.w[big::NLEN-1]>>TBITS;
            b.w[big::NLEN-1]&=TMASK;
            b.w[0]+=carry;
            
            b.w[(224/big::BASEBITS) as usize]+=carry<<(224%big::BASEBITS);
            b.norm();
            return b;
        }
       
        if MODTYPE==NOT_SPECIAL {
            let m = BIG::new_ints(&rom::MODULUS);              
            return BIG::monty(&m,rom::MCONST,d);
        }     
        return BIG::new();
    }

   /* convert to string */
	pub fn tostring(&mut self) -> String {
        let s=self.redc().tostring();
        return s;
    }

/* reduce this mod Modulus */
    pub fn reduce(&mut self) {
  		let p = BIG::new_ints(&rom::MODULUS);      	
        self.x.rmod(&p)
    }
    
/* test this=0? */
    pub fn iszilch(&mut self) -> bool {
        self.reduce();
        return self.x.iszilch();
    }
    
/* copy from FP b */
    pub fn copy(&mut self,b: &FP) {
        self.x.copy(&(b.x));
    }
    
/* copy from BIG b */
    pub fn bcopy(&mut self,b: &BIG) {
        self.x.copy(&b);
        self.nres();
    }

/* set this=0 */
    pub fn zero(&mut self) {
        self.x.zero();
    }
    
/* set this=1 */
    pub fn one(&mut self) {
        self.x.one(); self.nres()
    }
    
/* normalise this */
    pub fn norm(&mut self) {
        self.x.norm();
    }
/* swap FPs depending on d */
    pub fn cswap(&mut self,b: &mut FP,d: isize) {
        self.x.cswap(&mut (b.x),d);
    }
    
/* copy FPs depending on d */
    pub fn cmove(&mut self,b: &FP,d: isize) {
        self.x.cmove(&(b.x),d);
    }

/* this*=b mod Modulus */
    pub fn mul(&mut self,b: &FP)
    {
    //    self.norm();
    //    b.norm();
        if FP::pexceed(&(self.x),&(b.x)) {self.reduce()}

        let mut d=BIG::mul(&(self.x),&(b.x));
        self.x.copy(&FP::modulo(&mut d))
    }

    fn logb2(w: u32) -> usize {
        let mut v=w;
        v |= v >> 1;
        v |= v >> 2;
        v |= v >> 4;
        v |= v >> 8;
        v |= v >> 16;

        v = v - ((v >> 1) & 0x55555555);                 
        v = (v & 0x33333333) + ((v >> 2) & 0x33333333);  
        let r= ((   ((v + (v >> 4)) & 0xF0F0F0F)   * 0x1010101) >> 24) as usize;
        return r+1;    
    }

/* this = -this mod Modulus */
    pub fn neg(&mut self) {
  		let mut p = BIG::new_ints(&rom::MODULUS);   
    
    //    self.norm();

        let sb=FP::logb2((FP::excess(&(self.x))+1) as u32);

    //    let mut ov=BIG::excess(&(self.x));
    //    let mut sb=1; while ov != 0 {sb += 1;ov>>=1}
    
        p.fshl(sb);
        self.x.rsub(&p);
    
        if FP::excess(&(self.x))>=FEXCESS {self.reduce()}
    }

    /* this*=c mod Modulus, where c is a small int */
    pub fn imul(&mut self,c: isize) {
        let mut cc=c;
    //    self.norm();
        let mut s=false;
        if cc<0 {
            cc = -cc;
            s=true;
        }
        let afx=(FP::excess(&(self.x))+1)*((cc as Chunk)+1)+1;
        if cc<=big::NEXCESS && afx<FEXCESS {
            self.x.imul(cc);
            self.norm();
        } else {
            if afx<FEXCESS {
            	self.x.pmul(cc);
            } else {
  				let p = BIG::new_ints(&rom::MODULUS);               	
				let mut d=self.x.pxmul(cc);
				self.x.copy(&d.dmod(&p));
            }
        }
        if s {self.neg(); self.norm();}
    }

/* self*=self mod Modulus */
    pub fn sqr(&mut self) {
    //    self.norm();
        if FP::sexceed(&(self.x)) {self.reduce()}

        let mut d=BIG::sqr(&(self.x));
        self.x.copy(&FP::modulo(&mut d))
    }

/* self+=b */
    pub fn add(&mut self,b: &FP) {
        self.x.add(&(b.x));
        if FP::excess(&(self.x))+2>=FEXCESS {self.reduce()}
    }

/* self+=self */
    pub fn dbl(&mut self) {
        self.x.dbl();
        if FP::excess(&(self.x))+2>=FEXCESS {self.reduce()}
    }
    
/* self-=b */
    pub fn sub(&mut self,b: &FP)
    {
        let mut n=FP::new_copy(b);
        n.neg();
        self.add(&n);
    }    

/* self/=2 mod Modulus */
    pub fn div2(&mut self) {
    //    self.x.norm();
        if self.x.parity()==0 {
        	self.x.fshr(1);
        } else {
  			let p = BIG::new_ints(&rom::MODULUS);           	
            self.x.add(&p);
            self.x.norm();
            self.x.fshr(1);
        }
    }
/* self=1/self mod Modulus */
    pub fn inverse(&mut self) {
  		let mut p = BIG::new_ints(&rom::MODULUS);      	
        let mut r=self.redc();
        r.invmodp(&mut p);
        self.x.copy(&r);
        self.nres();
    }

/* return TRUE if self==a */
    pub fn equals(&mut self,a: &mut FP) -> bool {
        a.reduce();
        self.reduce();
        if BIG::comp(&(a.x),(&self.x))==0 {return true}
        return false;
    }   

/* return self^e mod Modulus */
    pub fn pow(&mut self,e: &mut BIG) -> FP {
      	let p = BIG::new_ints(&rom::MODULUS);   	
        let mut r=FP::new_int(1);
        e.norm();
        self.x.norm();
		let mut m=FP::new_copy(self);
        loop {
            let bt=e.parity();
            e.fshr(1);
            if bt==1 {r.mul(&m)}
            if e.iszilch() {break}
            m.sqr();
        }
        r.x.rmod(&p);
        return r;
    }

/* return sqrt(this) mod Modulus */
    pub fn sqrt(&mut self) -> FP {
        self.reduce();
      	let mut p = BIG::new_ints(&rom::MODULUS);  
        if MOD8==5 {
            p.dec(5); p.norm(); p.shr(3);
            let mut i=FP::new_copy(self); i.x.shl(1);
            let v=i.pow(&mut p);
            i.mul(&v); i.mul(&v);
            i.x.dec(1);
            let mut r=FP::new_copy(self);
            r.mul(&v); r.mul(&i);
            r.reduce();
            return r;
        }
        else
        {
            p.inc(1); p.norm(); p.shr(2);
            return self.pow(&mut p);
        }
    }
/* return jacobi symbol (this/Modulus) */
    pub fn jacobi(&mut self) -> isize
    {
     	let mut p = BIG::new_ints(&rom::MODULUS);      	
        let mut w=self.redc();
        return w.jacobi(&mut p);
    }

}
/*
fn main() {
    let p = BIG::new_ints(&rom::MODULUS);  
	let mut e = BIG::new_copy(&p);
	e.dec(1);

    let mut x = FP::new_int(3);
    let mut s=x.pow(&mut e);

    println!("s= {}",s.tostring());
}
*/

File: version3/swift/fp.swift
/*
	Licensed to the Apache Software Foundation (ASF) under one
	or more contributor license agreements.  See the NOTICE file
	distributed with this work for additional information
	regarding copyright ownership.  The ASF licenses this file
	to you under the Apache License, Version 2.0 (the
	"License"); you may not use this file except in compliance
	with the License.  You may obtain a copy of the License at
	
	http://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing,
	software distributed under the License is distributed on an
	"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
	KIND, either express or implied.  See the License for the
	specific language governing permissions and limitations
	under the License.
*/
//
//  fp.swift
//
//  Created by Michael Scott on 20/06/2015.
//  Copyright (c) 2015 Michael Scott. All rights reserved.
//  Small Finite Field arithmetic
//  AMCL mod p functions
//

final public class FP {


    static public let NOT_SPECIAL=0
    static public let PSEUDO_MERSENNE=1
    static public let MONTGOMERY_FRIENDLY=2
    static public let GENERALISED_MERSENNE=3

    static public let MODBITS:UInt = @NBT@
    static let MOD8:UInt = @M8@
    static public let MODTYPE =  @MT@   

    static let FEXCESS:Chunk = (1<<Chunk(BIG.BASEBITS*UInt(BIG.NLEN)-FP.MODBITS));
    static let OMASK:Chunk=Chunk(-1)<<Chunk(FP.MODBITS%BIG.BASEBITS)
    static let TBITS:UInt=FP.MODBITS%BIG.BASEBITS; // Number of active bits in top word
    static let TMASK:Chunk=(1<<Chunk(FP.TBITS))-1

    var x:BIG
    static let p=BIG(ROM.Modulus)

    static func EXCESS(_ a: BIG) -> Chunk
    {
        return ((a.w[BIG.NLEN-1] & FP.OMASK)>>Chunk(FP.MODBITS%BIG.BASEBITS))
    }

#if D32
    static func pexceed(_ a: BIG,_ b : BIG) -> Bool
    {
        let ea=EXCESS(a)
        let eb=EXCESS(b)
        if (DChunk(ea)+1)*(DChunk(eb)+1) > DChunk(FEXCESS) {return true}
        return false;
    }
    static func sexceed(_ a: BIG) -> Bool
    {
        let ea=EXCESS(a)
        if (DChunk(ea)+1)*(DChunk(ea)+1) > DChunk(FEXCESS) {return true}
        return false;
    }
#endif

#if D64
    static func pexceed(_ a: BIG,_ b : BIG) -> Bool
    {
        let ea=EXCESS(a)
        let eb=EXCESS(b)
        if (ea+1) > FEXCESS/(eb+1) {return true}
        return false;
    }
    static func sexceed(_ a: BIG) -> Bool
    {
        let ea=EXCESS(a)
        if (ea+1) > FEXCESS/(ea+1) {return true}
        return false;
    }
#endif

/* convert to Montgomery n-residue form */
    func nres()
    {
        if FP.MODTYPE != FP.PSEUDO_MERSENNE && FP.MODTYPE != FP.GENERALISED_MERSENNE
        {
            let d=DBIG(x)
            d.shl(UInt(BIG.NLEN)*BIG.BASEBITS)
            x.copy(d.mod(FP.p))
        }
    }
/* convert back to regular form */
    func redc() -> BIG
    {
        if FP.MODTYPE != FP.PSEUDO_MERSENNE && FP.MODTYPE != FP.GENERALISED_MERSENNE
        {
            let d=DBIG(x)
            return FP.mod(d)
        }
        else
        {
            let r=BIG(x)
            return r;
        }
    }
    
    /* reduce a DBIG to a BIG using the appropriate form of the modulus */
    static func mod(_ d: DBIG) -> BIG
    {
 
        if FP.MODTYPE==FP.PSEUDO_MERSENNE
        {
            let t=d.split(FP.MODBITS)
            var b=BIG(d)
            let v=t.pmul(Int(ROM.MConst))

    t.add(b)
    t.norm()


            let tw=t.w[BIG.NLEN-1]
            t.w[BIG.NLEN-1] &= TMASK
            t.inc(Int(ROM.MConst*((tw>>Chunk(FP.TBITS))+(v<<Chunk(BIG.BASEBITS-FP.TBITS)))))
    
    t.norm()
    return t

    //        b.add(t)
    //        b.norm()
    //        return b
        }
        if FP.MODTYPE==FP.MONTGOMERY_FRIENDLY
        {
            for i in 0 ..< BIG.NLEN {
                let (top,bot)=BIG.muladd(d.w[i],ROM.MConst-1,d.w[i],d.w[BIG.NLEN+i-1])
                d.w[BIG.NLEN+i]+=top; d.w[BIG.NLEN+i-1]=bot
 //                   d.w[BIG.NLEN+i]+=d.muladd(d.w[i],ROM.MConst-1,d.w[i],BIG.NLEN+i-1)
            }
    
            var b=BIG(0);
    
            for i in 0 ..< BIG.NLEN
            {
                b.w[i]=d.w[BIG.NLEN+i]
            }
            b.norm()
            return b;
        }
        if FP.MODTYPE==FP.GENERALISED_MERSENNE
        { // GoldiLocks Only
            let t=d.split(FP.MODBITS)
            let RM2=FP.MODBITS/2
            var b=BIG(d)
            b.add(t)
            let dd=DBIG(t)
            dd.shl(RM2)
            
            let tt=dd.split(FP.MODBITS)
            let lo=BIG(dd)
            b.add(tt)
            b.add(lo)
            b.norm()
            tt.shl(RM2)
            b.add(tt)
            
            let carry=b.w[BIG.NLEN-1]>>Chunk(FP.TBITS)
            b.w[BIG.NLEN-1]&=TMASK
            b.w[0]+=carry
            
            b.w[Int(224/BIG.BASEBITS)]+=carry<<Chunk(224%BIG.BASEBITS)
            b.norm()
            return b;
        }
        if FP.MODTYPE==FP.NOT_SPECIAL
        {
            let md=BIG(ROM.Modulus);

            return BIG.monty(md,ROM.MConst,d)
        }
        return BIG(0)
    }


    init()
    {
        x=BIG(0)
    }
    init(_ a: Int)
    {
        x=BIG(a)
        nres()
    }
    init(_ a: BIG)
    {
        x=BIG(a)
        nres()
    }
    init(_ a: FP)
    {
        x=BIG(a.x)
    }
    /* convert to string */
    func toString() -> String
    {
        let s=redc().toString();
        return s;
    }
    
    func toRawString() -> String
    {
        let s=x.toRawString();
        return s;
    }
/* reduce this mod Modulus */
    func reduce()
    {
        x.mod(FP.p)
    }
    
/* test this=0? */
    func iszilch() -> Bool
    {
        reduce();
        return x.iszilch()
    }
    
/* copy from FP b */
    func copy(_ b: FP)
    {
        x.copy(b.x);
    }
    
/* set this=0 */
    func zero()
    {
        x.zero();
    }
    
/* set this=1 */
    func one()
    {
        x.one(); nres()
    }
    
/* normalise this */
    func norm()
    {
        x.norm();
    }
/* swap FPs depending on d */
    func cswap(_ b: FP,_ d: Int)
    {
        x.cswap(b.x,d)
    }
    
/* copy FPs depending on d */
    func cmove(_ b: FP,_ d:Int)
    {
        x.cmove(b.x,d);
    }
/* this*=b mod Modulus */
    func mul(_ b: FP)
    {
    //    norm()
    //    b.norm()

        if FP.pexceed(x,b.x) {reduce()}

        //let ea=FP.EXCESS(x)
        //let eb=FP.EXCESS(b.x)
        
       // if Int64(ea+1)*Int64(eb+1)>Int64(FP.FEXCESS) {reduce()}
        /*if (ea+1)>=(FP.FEXCESS-1)/(eb+1) {reduce()}*/
        
        let d=BIG.mul(x,b.x)
        x.copy(FP.mod(d))
    }
    static func logb2(_ w: UInt32) -> Int
    {
        var v = w
        v |= (v >> 1)
        v |= (v >> 2)
        v |= (v >> 4)
        v |= (v >> 8)
        v |= (v >> 16)
        
        v = v - ((v >> 1) & 0x55555555)
        v = (v & 0x33333333) + ((v >> 2) & 0x33333333)
        let r = Int((   ((v + (v >> 4)) & 0xF0F0F0F)   &* 0x1010101) >> 24)
        return (r+1)
    }
    /* this = -this mod Modulus */
    func neg()
    {
        let m=BIG(FP.p);
    
    //    norm();
        let sb=FP.logb2(UInt32(FP.EXCESS(x))+1)
 //       var ov=BIG.EXCESS(x);
 //       var sb=1; while(ov != 0) {sb += 1;ov>>=1}
    
        m.fshl(sb)
        x.rsub(m)
    
        if FP.EXCESS(x)>=FP.FEXCESS {reduce()}
    }
    /* this*=c mod Modulus, where c is a small int */
    func imul(_ c: Int)
    {
        var cc=c
    //    norm();
        var s=false
        if (cc<0)
        {
            cc = -cc
            s=true
        }
        let afx=(FP.EXCESS(x)+1)*(cc+1)+1;
        if cc<=BIG.NEXCESS && afx<FP.FEXCESS
        {
            x.imul(cc)
            x.norm()
        }
        else
        {
            if afx<FP.FEXCESS {x.pmul(cc)}
            else
            {
				let d=x.pxmul(cc);
				x.copy(d.mod(FP.p));
            }
        }
        if s {neg();  norm()}
       
    }
    
/* this*=this mod Modulus */
    func sqr()
    {
    //    norm()

        if FP.sexceed(x) {reduce()}


     //   let ea=FP.EXCESS(x);
        
    //    if Int64(ea+1)*Int64(ea+1)>Int64(FP.FEXCESS) {reduce()}
        /*if (ea+1)>=(FP.FEXCESS-1)/(ea+1) {reduce()}*/
        
        let d=BIG.sqr(x);
        x.copy(FP.mod(d));
    }
    
    /* this+=b */
    func add(_ b: FP)
    {
        x.add(b.x);
        if FP.EXCESS(x)+2>=FP.FEXCESS {reduce()}
    }
/* this-=b */
    func sub(_ b: FP)
    {
        let n=FP(b)
        n.neg()
        self.add(n)
    }
/* this/=2 mod Modulus */
    func div2()
    {
    //    x.norm()
        if (x.parity()==0)
            {x.fshr(1)}
        else
        {
            x.add(FP.p)
            x.norm()
            x.fshr(1)
        }
    }
/* this=1/this mod Modulus */
    func inverse()
    {
        let r=redc()
        r.invmodp(FP.p)
        x.copy(r)
        nres()
    }
    
/* return TRUE if this==a */
    func equals(_ a: FP) -> Bool
    {
        a.reduce()
        reduce()
        if (BIG.comp(a.x,x)==0) {return true}
        return false;
    }
/* return this^e mod Modulus */
    func pow(_ e: BIG) -> FP
    {
        let r=FP(1)
        e.norm()
        x.norm()
	let m=FP(self)
        while (true)
        {
            let bt=e.parity()
            e.fshr(1)
            if bt==1 {r.mul(m)}
            if e.iszilch() {break}
            m.sqr();
        }
        r.x.mod(FP.p);
        return r;
    }
/* return sqrt(this) mod Modulus */
    func sqrt() -> FP
    {
        reduce();
        let b=BIG(FP.p)
        if (FP.MOD8==5)
        {
            b.dec(5); b.norm(); b.shr(3)
            let i=FP(self); i.x.shl(1)
            let v=i.pow(b)
            i.mul(v); i.mul(v)
            i.x.dec(1)
            let r=FP(self)
            r.mul(v); r.mul(i)
            r.reduce()
            return r
        }
        else
        {
            b.inc(1); b.norm(); b.shr(2)
            return pow(b)
        }
    }
/* return jacobi symbol (this/Modulus) */
    func jacobi() -> Int
    {
        let w=redc()
        return w.jacobi(FP.p)
    }
    
}
